contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+    def _check_overlap_predecessor(self):
+        for rec in self:
+            if rec.predecessor_contract_line_id:
+                if rec.date_start < rec.predecessor_contract_line_id.date_end:
+                    raise ValidationError(
+                        _("Contract line and its predecessor overlapped")
+                    )
+
+    @api.model
+    def _compute_first_recurring_next_date(
+        self,
+        date_start,
+        recurring_invoicing_type,
+        recurring_rule_type,
+        recurring_interval,
+    ):
+        if recurring_rule_type == 'monthlylastday':
+            return date_start + self.get_relative_delta(
+                recurring_rule_type, recurring_interval - 1
+            )
+        if recurring_invoicing_type == 'pre-paid':
+            return date_start
+        return date_start + self.get_relative_delta(
+            recurring_rule_type, recurring_interval
+        )
+
+    @api.onchange(
+        'is_auto_renew', 'auto_renew_rule_type', 'auto_renew_interval'
+    )
+    def _onchange_is_auto_renew(self):
+        """Date end should be auto-computed if a contract line is set to
+        auto_renew"""
+        for rec in self.filtered('is_auto_renew'):
+            rec.date_end = (
+                self.date_start
+                + self.get_relative_delta(
+                    rec.auto_renew_rule_type, rec.auto_renew_interval
+                )
+                - relativedelta(days=1)
+            )
+
+    @api.onchange(
+        'date_start',
+        'recurring_invoicing_type',
+        'recurring_rule_type',
+        'recurring_interval',
+    )
+    def _onchange_date_start(self):
+        for rec in self.filtered('date_start'):
+            rec.recurring_next_date = self._compute_first_recurring_next_date(
+                rec.date_start,
+                rec.recurring_invoicing_type,
+                rec.recurring_rule_type,
+                rec.recurring_interval,
+            )
+
+    @api.constrains('recurring_next_date', 'date_start')
+    def _check_recurring_next_date_start_date(self):
+        for line in self.filtered('recurring_next_date'):
+            if line.date_start and line.recurring_next_date:
+                if line.date_start > line.recurring_next_date:
+                    raise ValidationError(
+                        _(
+                            "You can't have a next invoicing date before the "
+                            "start of the contract '%s'"
+                        )
+                        % line.contract_id.name
+                    )
+
+    @api.constrains('recurring_next_date')
+    def _check_recurring_next_date_recurring_invoices(self):
+        for line in self.filtered('contract_id.recurring_invoices'):
+            if not line.recurring_next_date:
+                raise ValidationError(
+                    _(
+                        "You must supply a next invoicing date for contract "
```

Comment:
```suggestion
                        "You must supply a date of next invoice for contract "
```

(for consistency with the field name)